indexing-service/src/main/java/org/apache/druid/indexing/overlord/TaskRunner.java

```diff
@@ -88,11 +89,19 @@
    */
   void shutdown(String taskid, String reason);
 
-  default void shutdown(String taskid, String reasonFormat, Object... args)
+  default void shutdown(String taskid, String reasonFormat, Object... reasonArgs)
   {
-    shutdown(taskid, StringUtils.format(reasonFormat, args));
+    // only calculate the 'reason' string for debug level logging
+    // in large clusters the 'reasonArgs' may be very large / expensive if it includes a list of tasks
+    String reason = (getLogger().isDebugEnabled()) ? StringUtils.format(reasonFormat, reasonArgs) : "debug log disabled";
+    shutdown(taskid, reason);
   }
 
+  /**
+   * Get the logger. Not expected to be called by consumers.
+   */
+  Logger getLogger();
```

Comment:
I would probably change the default implementation to throw an `UnsupportedOperationException`. This will also have the benefit of not having to change classes implementing `TaskRunner` interface where they don't have a logger available (like test classes).